New trait.mcmc.Rdata by treatment

jessicaguo opened the pull request:
Function to be used in "workflow.R" after the meta-analysis step. Takes jagged.data and trait.mcmc to create a new "trait.mcmc.Rdata" object with posterior chains for each treatment. Target site/treatment is labeled "beta.o" for subsequent PEcAn steps. Former trait.mcmc object now saved as "trait.mcmc.original.Rdata"

Note: Requires that the pecan.xml has new entry such that settings$meta.analysis$treatment indicates desired treatment name, from BETYdb.

KristinaRiemer commented:
Just to make sure, this should be done in the `dbfiles` version of these, no the `pft` folder ones, due to how these are used downstream. See [this reference](https://github.com/danforthcenter/sentinel-detection/issues/189).

KristinaRiemer reviewed (changes requested):
So I just double checked and my earlier comment on this issue I think applies to this code. I think all of this should be done on the `trait.mcmc` and `jagged.data` in the `dbfiles/posterior/ID` folder instead of the `pft` folder. This should only require changing all of the files paths, because I'm prettttttty certain those two files are the same in the two places (though I don't really know why there are copies of them in the first place).

jessicaguo commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,90 @@
+  #save trt.match
+  save(trt.match, file = paste0(settings$outdir, "pft/SetariaWT_ME034/trt.match.Rdata"))
+  
+  #save existing trait.mcmc
```
Done

jessicaguo commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,90 @@
+  library(coda)
+  
+  #load jagged.data and trait.mcmc
+  load(paste0(settings$outdir, "pft/SetariaWT_ME034/jagged.data.Rdata"))
```
Done

jessicaguo commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,90 @@
+  
+  #load jagged.data and trait.mcmc
+  load(paste0(settings$outdir, "pft/SetariaWT_ME034/jagged.data.Rdata"))
+  load(paste0(settings$outdir, "pft/SetariaWT_ME034/trait.mcmc.Rdata"))
```
Done

jessicaguo commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,90 @@
+  trt.match <- lapply(jagged.data, collapse)[names(trait.mcmc)]
+  
+  #save trt.match
+  save(trt.match, file = paste0(settings$outdir, "pft/SetariaWT_ME034/trt.match.Rdata"))
```
Done

jessicaguo commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,90 @@
+  cnames <- colnames(mc[[1]])
+  for(i in 1:nrow(trt)){
+    for(j in 1:ncol(trt)){
+      col_ind[i,j] <- if(length(which(cnames == paste0("beta.", colnames(trt)[j], "[", trt[i,j], "]"))) == 1){
```
Great suggestion!

dlebauer commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,98 @@
+  library(coda)
+  
+  #load jagged.data and trait.mcmc
+  load(paste0(settings$outdir, "/pft/SetariaWT_ME034/jagged.data.Rdata"))
```
- use `file.path` instead of paste0
- use `settings$pft$name` instead of `SetariaWT_ME034`

dlebauer commented on BioCro/DARPA/set.MA.trt.R:
```diff
@@ -0,0 +1,98 @@
+  
+  #create new trait.mcmc of combined random effects
+  new.trait.mcmc <- mapply(FUN = RE.combine, mc = trait.mcmc, trt = trt.match, 
+                           SIMPLIFY = F)
```
always use `TRUE` and `FALSE` instead of `T` and `F`

jessicaguo commented:
Updated to better reflect the tidyverse style guide. Following PEcAn convention, used file.path() instead of paste0() to read in and write out files. Used existing PEcAn function dbfile.file() to obtain the correct dbfiles/posterior folder; outputs now saved in both dbfiles and pft. Fixed typo that was accidentally saving out trt.match as 3 different objects.

KristinaRiemer commented on BioCro/DARPA/set_MA_trt.R:
```diff
@@ -0,0 +1,108 @@
+  #index of target treatment
+  ind <- which(trt$site_id == target.site & trt$trt_name == target.trt)
+  
+  #for each chain, add trt_name as column names; then sub out the traget treatment column name with "beta.o"
```
```suggestion
  #for each chain, add treatment names as column names; then sub out the target treatment column name with "beta.o"
```

KristinaRiemer commented on BioCro/DARPA/set_MA_trt.R:
```diff
@@ -0,0 +1,108 @@
+  
+  #for each chain, add trt_name as column names; then sub out the traget treatment column name with "beta.o"
+  for(c in 1:length(trait.mcmc)){
+    colnames(trait.mcmc[[c]]) <- trt$trt_name
```
Are these for sure always going to be in the same order?

jessicaguo replied:
See line 99, where ind is evaluated as the correct column index.

KristinaRiemer commented on BioCro/DARPA/set_MA_trt.R:
```diff
@@ -0,0 +1,108 @@
+                             SIMPLIFY = FALSE)
+  
+  #rename as trait.mcmc and save as "trait.mcmc.Rdata", 2 places
+  trait.mcmc <- final.trait.mcmc
```
This renaming step seems possibly unnecessary? `final.trait.mcmc` could just directly be saved as "trait.mcmc.Rdata" in lines 47 and 48.

jessicaguo replied:
Hmm yes, but then when loaded in for the next step, it will be named final.trait.mcmc. I thought maybe the next function would load the Rdata object and then assume it was named trait.mcmc

KristinaRiemer replied:
Yeah, you're right about that, the .Rdata file needs to be named trait.mcmc. But line 47 could be `save(final.trait.mcmc, file = file.path(settings$database$dbfiles, "posterior", postid, "trait.mcmc.Rdata"))`, right. That would then make line 46 unnecessary.

jessicaguo replied:
When the .Rdata object loads, it will retain the object name it was saved as (I'm pretty sure). So then the next step will load in the file name "trait.mcmc.Rdata", but the object name will still be "final.trait.mcmc". So if the function requires the object name to be trati.mcmc, it still won't locate it.

KristinaRiemer replied:
Ah, okay. I'm not as familiar with .Rdata files, so some of their behavior seems odd to me.

KristinaRiemer reviewed (approved):
This works great! Thanks so much for doing all of this @jessicaguo!
